### PR TITLE
Typed Variable Plugin

### PR DESCRIPTION
**Pull request by alschmiedt**
Guide to this very large PR (sorry): 
Main Files: 
Modal.js - I pulled out the main functionality for a modal, so if we want we can pull this into its own plugin. At the very least if we want to do something similar again we will have a n easy starting point.
TypedModal.js - Extends Modal and adds features necessary for a typed modal. 

Test Folder:
InputModal - Just my way of checking to see how easy it was to change
multi-playgorund.html
playground.html

Still In Progress: 
1. Tests
2. Centering on the workspace - I want to add this on the modal but what I need isn't in metrics so still need to figure out how to get what I need from Blockly.

Note: If you want to test multiple workspaces you will need to build the blockly.min.js file and then copy it into the node_modules folder in test, since this depends on the latest version of Blockly.
Screen Shot 2020-03-25 at 3 12 20 PM

### CODE REVIEW

File: packages/plugins/typed-modal/package.json [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    "typed variables"
+  ],
+  "homepage": "https://github.com/google/blockly-samples/tree/master/packages/plugins/typed-modal#readme",
+  "bugs": {
+    "url": "https://github.com/google/blockly-samples/issues"
+  },
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/google/blockly-samples.git",
+    "directory": "packages/plugins/blockly-typed-modal"
```
Comment (samelhusseini) [inline packages/plugins/typed-modal/package.json]: packages/plugins/typed-modal

File: packages/plugins/typed-modal/src/Modal.js [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+    this.addEvent_(/** @type{!HTMLDivElement} */ this.htmlDiv_, 'keydown',
+        this, this.handleKeyDown_);
+  }
+
+  /**
+   * Inject the css for a Blockly modal.
+   * @protected
+   */
+  injectCss_() {
+    injectCss('blockly-modal-css', `
```
Comment (samelhusseini) [inline packages/plugins/typed-modal/src/Modal.js]: I still think there's little reason not to use Blockly.CSS.register. We're shipping css injection logic in every plugin and that doesn't make sense.

File: packages/plugins/typed-modal/src/Modal.js [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+      this.htmlDiv_.remove();
+      this.htmlDiv_ = null;
+    }
+  }
+
+  /**
+   * Shows the Blockly modal and focus on the first focusable element.
+   */
+  show() {
+    this.focusableEls = this.htmlDiv_.querySelectorAll('a[href],' +
```
Comment (samelhusseini) [inline packages/plugins/typed-modal/src/Modal.js]: Test on IE, querySelectorAll doesn't have full support there.

File: packages/plugins/typed-modal/src/Modal.js [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+   */
+  handleKeyDown_(e) {
+    if (e.keyCode === Blockly.utils.KeyCodes.TAB) {
+      if (this.focusableEls.length <= 1) {
+        e.preventDefault();
+        e.stopPropagation();
+        return;
+      }
+      if (e.shiftKey) {
+        this.handleBackwardTab_(e);
```
Comment (samelhusseini) [inline packages/plugins/typed-modal/src/Modal.js]: Seems like we're doing more work than we need to here. You're taking a snapshot of all the focusable elements, and then handling tab navigation on each and every one of them, rather: 

Couldn't you just handle "wrap" tab navigation for the first and last element, to tab lock and ensure the user doesn't end up outside of the modal.

Good resource for implementing a dialog here: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/dialog_role

File: packages/plugins/typed-modal/package.json (hunk 1 above)
Comment (alschmiedt) [reply packages/plugins/typed-modal/package.json]: done

File: packages/plugins/typed-modal/src/Modal.js (hunk 2 above)
Comment (alschmiedt) [reply packages/plugins/typed-modal/src/Modal.js]: I did try to use Blockly.Css.Register, but I get a "CSS already injected" error. It looks like because the css is injected during createDom and the modal is created after this. However,I agree it is weird to keep including. Will look into alternative solution.

File: packages/plugins/typed-modal/src/Modal.js (hunk 3 above)
Comment (alschmiedt) [reply packages/plugins/typed-modal/src/Modal.js]: I did test and it seems to be working. I think the only bug I found was if there is nothing in the modal it throws an error. However all modals should have at least the x in the corner so seems like a corner case.

File: packages/plugins/typed-modal/src/Modal.js (hunk 4 above)
Comment (alschmiedt) [reply packages/plugins/typed-modal/src/Modal.js]: I think that is what I am doing. The handleForwardTab and handlBackwardTab only check if I need to wrap, if I don't then they don't do anything. 
Looking at it more I don't need to hold all the focusableEls so updated to have a firstFocusableEl and lastFocusableEl

File: packages/plugins/typed-modal/.eslintrc.json [hunk 5]
```diff
@@ -0,0 +1,99 @@
+{
```
Comment (rachel-fenichel) [inline packages/plugins/typed-modal/.eslintrc.json]: Can you move this up to `packages/plugins/` so that you don't have to recreate it for each plugin?

File: packages/plugins/typed-modal/src/Modal.js (hunk 2 above)
Comment (samelhusseini) [reply packages/plugins/typed-modal/src/Modal.js]: You should be able to add the CSS to be registered during page load, rather than during modal initialization, like the toolbox in Blockly for example: 
https://github.com/google/blockly/blob/master/core/toolbox.js#L741

File: packages/plugins/typed-modal/.eslintrc.json (hunk 5 above)
Comment (alschmiedt) [reply packages/plugins/typed-modal/.eslintrc.json]: done

File: packages/plugins/typed-modal/src/Modal.js [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+     * HTML container for the modal.
+     * @type {HTMLDivElement}
+     * @private
+     */
+    this.htmlDiv_ = null;
+
+    /**
+     * Array holding info needed to unbind events.
+     * Used for disposing.
+     * Ex: [[node, name, func], [node, name, func]].
```
Comment (rachel-fenichel) [inline packages/plugins/typed-modal/src/Modal.js]: API thought: should we expose `Blockly.boundEvent` or similar?

File: packages/plugins/typed-modal/src/TypedModal.js [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+   * Dispose of the typed modal.
+   * @override
+   */
+  dispose() {
+    super.dispose();
+    this.workspace_.removeButtonCallback(this.btnCallBackName_);
+    this.variableNameInput_ = null;
+    this.variableTypesDiv_ = null;
+    this.firstTypeInput_ = null;
+    // TODO: Remove style tag.
```
Comment (rachel-fenichel) [inline packages/plugins/typed-modal/src/TypedModal.js]: Remove todo

File: packages/plugins/typed-modal/src/TypedModal.js [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+  }
+
+  /**
+   * Dispose of the typed modal.
+   * @override
+   */
+  dispose() {
+    super.dispose();
+    this.workspace_.removeButtonCallback(this.btnCallBackName_);
+    this.variableNameInput_ = null;
```
Comment (rachel-fenichel) [inline packages/plugins/typed-modal/src/TypedModal.js]: Explicitly setting these to null is an old blockly pattern, but I don't think you need it for these--the garbage collector will get them--and we're phasing it out in blockly more generally.

**Review by rachel-fenichel — COMMENTED** [review]
Looks pretty solid overall.  A naming nit: I think it should be TypedVariableModal, not TypedModal, so that the adjective is modifying the correct thing.

File: packages/plugins/typed-modal/src/TypedModal.js (hunk 7 above)
Comment (alschmiedt) [reply packages/plugins/typed-modal/src/TypedModal.js]: done

File: packages/plugins/typed-modal/src/TypedModal.js (hunk 8 above)
Comment (alschmiedt) [reply packages/plugins/typed-modal/src/TypedModal.js]: done

File: packages/plugins/typed-variable-modal/src/TypedVariableModal.js [hunk 9]
```diff
@@ -0,0 +1,368 @@
...
+   * @param {string} btnCallbackName The name used to register the button
+   *     callback.
+   * @param {Array<Array<string>>} types An array holding arrays with the
+   *     display name as the first value and the type as the second.
+   *     Ex: [['Penguin', 'PENGUIN'], ['Giraffe', 'GIRAFFE']].
+   * @param {TypedVarModalMessages} opt_messages The messages for a typed variable
+   *     modal.
+   */
+  constructor(workspace, btnCallbackName, types, opt_messages) {
+    const title = opt_messages ? opt_messages["TYPED_VAR_MODAL_TITLE"] :
```
Comment (samelhusseini) [inline packages/plugins/typed-variable-modal/src/TypedVariableModal.js]: Why don't we use Blockly.Msg to house these instead?

File: packages/plugins/typed-modal/src/Modal.js (hunk 6 above)
Comment (alschmiedt) [reply packages/plugins/typed-modal/src/Modal.js]: We use it in both plugins so far, and I can see us using it in future plugins so it would probably be nice to not have to keep duplicating.

File: packages/plugins/typed-variable-modal/test/InputModal.js [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    this.firstTypeInput_ = selectDiv.querySelector('.input-modal-type');
+    return selectDiv;
+  }
+
+  /**
+   * Get the selected type.
+   * @return {string} The selected type.
+   */
+  getSelectedType() {
+    const selectDiv = this.variableTypesDiv_;
```
Comment (samelhusseini) [inline packages/plugins/typed-variable-modal/test/InputModal.js:32]: What's the use case for this?

File: packages/plugins/typed-variable-modal/webpack.config.js [hunk 11]
```diff
@@ -0,0 +1,52 @@
+/**
```
Comment (samelhusseini) [inline packages/plugins/typed-variable-modal/webpack.config.js]: You can now use ``../../webpack.config.js``.

File: packages/plugins/typed-variable-modal/src/Modal.js [hunk 12]
```diff
@@ -0,0 +1,368 @@
...
+    this.htmlDiv_.setAttribute('role', 'dialog');
+    // End creating the container
+
+    // Create the header
+    const modalHeader = document.createElement('header');
+    Blockly.utils.dom.addClass(modalHeader, 'blockly-modal-header');
+
+    this.renderHeader_(modalHeader);
+
+    const exitButton = document.createElement('button');
```
Comment (samelhusseini) [inline packages/plugins/typed-variable-modal/src/Modal.js:275]: It's a best practice to place the exit button at the bottom of the DOM, otherwise its always going to be the first focusable element (which isnt a great user experience for a user using a screen reader)

File: packages/plugins/typed-variable-modal/test/InputModal.js (hunk 10 above)
Comment (alschmiedt) [reply packages/plugins/typed-variable-modal/test/InputModal.js:32]: In the InputModal I use an Input instead of a list of types to get the selected type. I needed a  way to give the selectedType. Probably doesn't need to be public though, so I can change that.

Comment (samelhusseini) [reply packages/plugins/typed-variable-modal/test/InputModal.js:32]: Yeah, I just don't see a use case for this outside of the 'protected' case.

File: packages/plugins/typed-variable-modal/webpack.config.js (hunk 11 above)
Comment (alschmiedt) [reply packages/plugins/typed-variable-modal/webpack.config.js]: done

File: packages/plugins/typed-variable-modal/src/Modal.js (hunk 12 above)
Comment (alschmiedt) [reply packages/plugins/typed-variable-modal/src/Modal.js:275]: Ok, I removed the X in the top right corner and changed the cancel button to be in the modal.

Comment (samelhusseini) [reply packages/plugins/typed-variable-modal/src/Modal.js:275]: Curious why you decided to remove it?

File: packages/plugins/typed-variable-modal/src/TypedVariableModal.js (hunk 9 above)
Comment (alschmiedt) [reply packages/plugins/typed-variable-modal/src/TypedVariableModal.js]: I think the idea was to make it standalone. We don't necessarily know what Blockly is going to do with its messages in the future and by making this separate it gives us the benefit of not having to update in the future. Also, because we are not planning on translating any of these messages I don't see the benefit we get from having it on Blockly.Msg.

Comment (samelhusseini) [reply packages/plugins/typed-variable-modal/src/TypedVariableModal.js]: Fair enough.
If Blockly changes messages in the future, I would imagine we would continue to be backwards compatible with Blockly.Msg. 
The pro of using Blockly.Msg that I think you're missing with this is it makes it so the plugin is compatible with a developer's localization flow. 

Blockly uses Blockly.Msg to house localized messages, internally, but if a developer has additional blocks, fields, etc and wants to localize those, they would do so with their own localization flow (translatewiki, crowdin, etc), and the output of these processes is a JSON that is dumped into Blockly.Msg. In other words, if they add these additional messages to their flow, everything would automagically work if you use Blockly.Msg.

Comment (alschmiedt) [reply packages/plugins/typed-variable-modal/src/TypedVariableModal.js]: Ok that makes sense. I changed so that the typed modal uses Blockly.Msg.

File: packages/plugins/typed-variable-modal/src/Modal.js (hunk 12 above)
Comment (alschmiedt) [reply packages/plugins/typed-variable-modal/src/Modal.js:275]: Talked about offline a bit. In the end I put the 'x' button back and kept it in the header. When I show the modal I skip over the 'x' button and focus the next item in the list if one exists. The reason I didn't position absolutely is because I didn't have all the positioning info I needed. Also, if I positioned absolutely and someone had a very long title then the button would overlap with it.

File: packages/plugins/typed-variable-modal/README.md [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+
+Create your workspace.
+```js
+workspace = Blockly.inject('blocklyDiv', {
+    toolbox: toolbox
+});
+```
+
+Add a custom category to your toolbox.
+```js
```
Comment (samelhusseini) [inline packages/plugins/typed-variable-modal/README.md]: xml

File: packages/plugins/typed-variable-modal/package.json [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+  "name": "@blockly/plugin-typed-var-modal",
+  "version": "1.0.0",
+  "description": "A Blockly plugin to create a modal for creating typed variables.",
+  "author": "Blockly Team",
+  "keywords": [
+    "blockly",
+    "modal",
+    "typed variables"
+  ],
+  "homepage": "https://github.com/google/blockly-samples/tree/master/packages/plugins/typed-var-modal#readme",
```
Comment (samelhusseini) [inline packages/plugins/typed-variable-modal/package.json]: rename to new directory name: typed-variable-modal

File: packages/plugins/typed-variable-modal/package.json [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+  "bugs": {
+    "url": "https://github.com/google/blockly-samples/issues"
+  },
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/google/blockly-samples.git",
+    "directory": "packages/plugins/typed-var-modal"
```
Comment (samelhusseini) [inline packages/plugins/typed-variable-modal/package.json]: same here: typed-variable-modal

File: packages/plugins/typed-variable-modal/src/Modal.js [hunk 16]
```diff
@@ -0,0 +1,372 @@
...
+     * @private
+     */
+    this.boundEvents_ = [];
+
+    /**
+     * If true close the modal when the user clicks outside the modal.
+     * Otherwise, only close when user hits the 'X' button or escape.
+     * @type {boolean}
+     */
+    this.closeOnClick = true;
```
Comment (samelhusseini) [inline packages/plugins/typed-variable-modal/src/Modal.js]: Maybe rename to shouldCloseOnOverlayClick?

Comment (samelhusseini) [reply packages/plugins/typed-variable-modal/src/Modal.js]: Also, should we add one for shouldCloseOnEsc?

File: packages/plugins/typed-variable-modal/src/Modal.js [hunk 17]
```diff
@@ -0,0 +1,372 @@
...
+    this.htmlDiv_.appendChild(modalFooter);
+  }
+
+  /**
+   * Render content for the modal header.
+   * @param {HTMLElement} headerContainer The modal's header div.
+   * @protected
+   */
+  renderHeader_(headerContainer) {
+    const modalTitle = document.createElement('H2');
```
Comment (samelhusseini) [inline packages/plugins/typed-variable-modal/src/Modal.js]: nit: lowercase h2.

File: packages/plugins/typed-variable-modal/src/Modal.js [hunk 18]
```diff
@@ -0,0 +1,372 @@
...
+   * Render content for the modal footer.
+   * @param {HTMLElement} _footerContainer The modal's footer div.
+   * @protected
+   */
+  renderFooter_(_footerContainer) {
+    // No-op on the base class.
+  }
+}
+Blockly.Css.register([`
+    .blockly-modal-container {
```
Comment (samelhusseini) [inline packages/plugins/typed-variable-modal/src/Modal.js]: I'm seeing different styles of CSS here. In blockly we use camelcase, and not `-`. Any reason to not stick with the blockly style?

File: packages/plugins/typed-variable-modal/src/TypedVariableModal.js [hunk 19]
```diff
@@ -0,0 +1,385 @@
...
+
+    /**
+     * An array holding arrays with the name of the type and the display name
+     *     for the type. Ex: [['Penguin', 'PENGUIN'], ['Giraffe', 'GIRAFFE']].
+     * @type {Array<Array<string>>}
+     * @protected
+     */
+    this.types_ = types;
+
+    const messages = opt_messages || {
```
Comment (samelhusseini) [inline packages/plugins/typed-variable-modal/src/TypedVariableModal.js]: Might be better to do something like. 
Blockly.utils.object.mixin({
 "TYPED_VAR_MODAL_TITLE": "... ",
 ... 
}, opt_messages);

That way you ensure that we have at least the EN messages when querying Blockly.Msg['TYPED_VAR...']

File: packages/plugins/typed-variable-modal/src/TypedVariableModal.js [hunk 20]
```diff
@@ -0,0 +1,385 @@
...
+   * @param {string} btnCallbackName The name used to register the button
+   *     callback.
+   * @param {Array<Array<string>>} types An array holding arrays with the
+   *     display name as the first value and the type as the second.
+   *     Ex: [['Penguin', 'PENGUIN'], ['Giraffe', 'GIRAFFE']].
+   * @param {TypedVarModalMessages} opt_messages The messages for a typed variable
+   *     modal.
+   */
+  constructor(workspace, btnCallbackName, types, opt_messages) {
+    const title = opt_messages ? opt_messages["TYPED_VAR_MODAL_TITLE"] :
```
Comment (samelhusseini) [inline packages/plugins/typed-variable-modal/src/TypedVariableModal.js]: If you're passed in a opt_messages bag that doesn't contain 'TYPED_VAR_MODAL_TITLE', you get no title. 

Alternately you could do: 
```
(opt_messages && opt_messages['TYPED_VAR_MODAL_TITLE']) || 'Create Typed Variable';
```

File: packages/plugins/typed-variable-modal/src/TypedVariableModal.js [hunk 21]
```diff
@@ -0,0 +1,385 @@
...
+  }
+
+  /**
+   * Set the messages for the typed variable modal.
+   * Used to change the location.
+   * @param {!TypedVarModalMessages} messages The messages needed to create a
+   *     typed modal.
+   */
+  setLocale(messages) {
+    Object.keys(messages).forEach(function(k) {
```
Comment (samelhusseini) [inline packages/plugins/typed-variable-modal/src/TypedVariableModal.js]: nit: Use arrow function.

File: packages/plugins/typed-variable-modal/src/TypedVariableModal.js [hunk 22]
```diff
@@ -0,0 +1,385 @@
...
+   * @override
+   */
+  renderContent_(contentContainer) {
+    const varNameContainer = this.createVarNameContainer_();
+    this.variableNameInput_ = varNameContainer
+        .querySelector('.typed-modal-variable-name-input');
+
+    const typedVarDiv = document.createElement('div');
+    Blockly.utils.dom.addClass(typedVarDiv, 'typed-modal-types');
+    typedVarDiv.innerText = Blockly.Msg["TYPED_VAR_MODAL_TYPES_LABEL"];
```
Comment (samelhusseini) [inline packages/plugins/typed-variable-modal/src/TypedVariableModal.js]: use textContent instead.

File: packages/plugins/typed-variable-modal/src/TypedVariableModal.js [hunk 23]
```diff
@@ -0,0 +1,385 @@
...
+   * @override
+   */
+  renderFooter_(footerContainer) {
+    footerContainer.appendChild(this.createConfirmBtn_());
+    footerContainer.appendChild(this.createCancelBtn_());
+  }
+
+  /**
+   * Create button in charge of creating the variable.
+   * @return {!HTMLButtonElement}The button in charge of creating a variable.
```
Comment (samelhusseini) [inline packages/plugins/typed-variable-modal/src/TypedVariableModal.js]: nit: extra space after }

File: packages/plugins/typed-variable-modal/src/TypedVariableModal.js [hunk 24]
```diff
@@ -0,0 +1,385 @@
...
+   * @protected
+   */
+  createConfirmBtn_() {
+    const confirmBtn = document.createElement('button');
+    Blockly.utils.dom.addClass(confirmBtn, 'blockly-modal-btn');
```
Comment (samelhusseini) [inline packages/plugins/typed-variable-modal/src/TypedVariableModal.js]: I would avoid using ``Blockly.utils.dom.addClass`` unless you have a need to add / remove a class. In this case you're creating the element right there, simply setting ``className`` to ``blockly-modal-btn  blockly-modal-btn-primary`` is clearer.

File: packages/plugins/typed-variable-modal/src/TypedVariableModal.js [hunk 25]
```diff
@@ -0,0 +1,385 @@
...
+    Blockly.utils.dom.addClass(confirmBtn, 'blockly-modal-btn-primary');
+    confirmBtn.innerText = Blockly.Msg['TYPED_VAR_MODAL_CONFIRM_BUTTON'];
```
Comment (samelhusseini) [inline packages/plugins/typed-variable-modal/src/TypedVariableModal.js]: In general go with textContent. It doesn't have security implications like innerHTML (security), and doesn't parse HTML like innerText (performance).

File: packages/plugins/typed-variable-modal/src/TypedVariableModal.js [hunk 26]
```diff
@@ -0,0 +1,385 @@
...
+
+  /**
+   * Create button in charge of cancelling.
+   * @return {!HTMLButtonElement} The button in charge of cancelling.
+   * @protected
+   */
+  createCancelBtn_() {
+    const cancelBtn = document.createElement('button');
+    Blockly.utils.dom.addClass(cancelBtn, 'blockly-modal-btn');
+    cancelBtn.innerText = Blockly.Msg['TYPED_VAR_MODAL_CANCEL_BUTTON'];
```
Comment (samelhusseini) [inline packages/plugins/typed-variable-modal/src/TypedVariableModal.js]: Same as above. Use className = '...', and textContent

File: packages/plugins/typed-variable-modal/test/InputModal.js [hunk 27]
```diff
@@ -0,0 +1,64 @@
...
+
+  /**
+   * @override
+   */
+  renderFooter_(footerContainer) {
+    footerContainer.appendChild(this.createConfirmBtn_());
+    const randomBtn = document.createElement('button');
+    Blockly.utils.dom.addClass(randomBtn, 'blockly-modal-btn');
+    randomBtn.innerText = 'Create random name';
+    this.addEvent_(randomBtn,'click',this, ()=> this.randomName());
```
Comment (samelhusseini) [inline packages/plugins/typed-variable-modal/test/InputModal.js]: nit, spacing after 'click'

File: packages/plugins/typed-variable-modal/README.md (hunk 13 above)
Comment (alschmiedt) [reply packages/plugins/typed-variable-modal/README.md]: done

File: packages/plugins/typed-variable-modal/package.json (hunk 14 above)
Comment (alschmiedt) [reply packages/plugins/typed-variable-modal/package.json]: done

File: packages/plugins/typed-variable-modal/package.json (hunk 15 above)
Comment (alschmiedt) [reply packages/plugins/typed-variable-modal/package.json]: done

File: packages/plugins/typed-variable-modal/src/Modal.js (hunk 17 above)
Comment (alschmiedt) [reply packages/plugins/typed-variable-modal/src/Modal.js]: done

File: packages/plugins/typed-variable-modal/src/TypedVariableModal.js (hunk 20 above)
Comment (alschmiedt) [reply packages/plugins/typed-variable-modal/src/TypedVariableModal.js]: done

File: packages/plugins/typed-variable-modal/src/TypedVariableModal.js (hunk 19 above)
Comment (alschmiedt) [reply packages/plugins/typed-variable-modal/src/TypedVariableModal.js]: done

File: packages/plugins/typed-variable-modal/src/TypedVariableModal.js (hunk 21 above)
Comment (alschmiedt) [reply packages/plugins/typed-variable-modal/src/TypedVariableModal.js]: done

File: packages/plugins/typed-variable-modal/src/TypedVariableModal.js (hunk 22 above)
Comment (alschmiedt) [reply packages/plugins/typed-variable-modal/src/TypedVariableModal.js]: done

File: packages/plugins/typed-variable-modal/src/TypedVariableModal.js (hunk 23 above)
Comment (alschmiedt) [reply packages/plugins/typed-variable-modal/src/TypedVariableModal.js]: done

File: packages/plugins/typed-variable-modal/src/TypedVariableModal.js (hunk 24 above)
Comment (alschmiedt) [reply packages/plugins/typed-variable-modal/src/TypedVariableModal.js]: done

File: packages/plugins/typed-variable-modal/src/TypedVariableModal.js (hunk 25 above)
Comment (alschmiedt) [reply packages/plugins/typed-variable-modal/src/TypedVariableModal.js]: done

File: packages/plugins/typed-variable-modal/src/TypedVariableModal.js (hunk 26 above)
Comment (alschmiedt) [reply packages/plugins/typed-variable-modal/src/TypedVariableModal.js]: done

File: packages/plugins/typed-variable-modal/test/InputModal.js (hunk 27 above)
Comment (alschmiedt) [reply packages/plugins/typed-variable-modal/test/InputModal.js]: done

**Review by samelhusseini — APPROVED** [review]
Overall looks good, a few nits and bits worth attending to.

File: packages/plugins/typed-variable-modal/src/Modal.js (hunk 16 above)
Comment (alschmiedt) [reply packages/plugins/typed-variable-modal/src/Modal.js]: Added both

File: packages/plugins/typed-variable-modal/src/Modal.js (hunk 18 above)
Comment (alschmiedt) [reply packages/plugins/typed-variable-modal/src/Modal.js]: We use the - in workspace search plugin. It seems like that is the most agreed upon style for css naming in the community and what the google style guide says as well. I am fine either way, but would prefer that we keep all the plugins the same. Depending on what we decide will update workspace search to match.

Comment (samelhusseini) [reply packages/plugins/typed-variable-modal/src/Modal.js]: Agreed on consistency, and I'm okay with diverging from blockly style guide (to google style guide) if we're consistent about it. Problem here is things like field_date and field_slider would have to change (only because they live in blockly-samples), and particularly for fields, some of the field CSS will follow one pattern and others a different pattern only because it lives in a different repo. 

I would recommend we follow the blockly style guide for things that exist in Blockly. Obviously for ES6 style things, we would have to match the Google style guide.

File: packages/plugins/typed-variable-modal/src/Modal.js [hunk 28]
```diff
@@ -0,0 +1,378 @@
...
+    this.boundEvents_.push(event);
+  }
+
+  /**
+   * Create all the dom elements for the modal.
+   */
+  render() {
+    /*
+     * Creates the Modal. The generated modal looks like:
+     * <div class="blockly-modal-container" role="dialog">
```
Comment (samelhusseini) [inline packages/plugins/typed-variable-modal/src/Modal.js]: I know that we've split up the "content" of the modal into a header / content / footer, but these are just semantic divisions. 
The "content" of the modal (ie: the div that gets styled to be in the center) in this case is the container div, and i don't see a div for the overlay (in case you wanted to style that differently). 

Could we have an overlay div and a content div.
```
<div class="blocklyModalOverlay">
  <div class="blocklyModalContent">
      <header class="blocklyModalHeader"></header>
      <main class="blocklyModalMainContent"></main>
      <footer class="blocklyModalHeader"></footer>
  </div>
</div>
```
or similar.

This lets you style the modal like this: 
Screen Shot 2020-04-01 at 2 43 50 PM

Comment (alschmiedt) [reply packages/plugins/typed-variable-modal/src/Modal.js]: I think the "overlay" is the widget div. When we show the modal we add it into the widget div. I could add a class to the widget div so they could do something similar to that?

Comment (samelhusseini) [reply packages/plugins/typed-variable-modal/src/Modal.js]: (In summary), there's no way to style the overlay.

Comment (samelhusseini) [reply packages/plugins/typed-variable-modal/src/Modal.js]: You could add a style to the widget div, but I think and I'm seeing this in the widgetDispose, that's adding a lot of extra complexity on widget cleanup.
Why not just add another div, say blocklyModalOverlay and set all event handlers, and styling on that.

Making the widget div just a container for housing things, that is simply cleared on hideChaff. The other benefit of that, is you wouldn't have to compete with .blocklyWidgetDiv CSS elsewhere in the editor for what you're trying to achieve.

File: packages/plugins/typed-variable-modal/src/Modal.js [hunk 29]
```diff
@@ -0,0 +1,378 @@
...
+   */
+  renderFooter_(_footerContainer) {
+    // No-op on the base class.
+  }
+}
+Blockly.Css.register([`
+    .blockly-modal-container {
+      background-color: white;
+      border: 1px solid gray;
+      max-width: 50%;
```
Comment (samelhusseini) [inline packages/plugins/typed-variable-modal/src/Modal.js]: why set max-width to 50%, and then width to 400px. That doesn't make sense to me?

Comment (alschmiedt) [reply packages/plugins/typed-variable-modal/src/Modal.js]: whoops, left over from a previous version. Removed.

File: packages/plugins/typed-variable-modal/src/Modal.js (hunk 28 above)
Comment (alschmiedt) [reply packages/plugins/typed-variable-modal/src/Modal.js]: Ok updated.

File: packages/plugins/typed-variable-modal/src/Modal.js [hunk 30]
```diff
@@ -0,0 +1,381 @@
...
+  /**
+   * Render content for the modal header.
+   * @param {HTMLElement} headerContainer The modal's header div.
+   * @protected
+   */
+  renderHeader_(headerContainer) {
+    const modalTitle = document.createElement('h2');
+    modalTitle.className = 'blocklyModalHeaderTitle';
+    modalTitle.appendChild(document.createTextNode(this.title_));
+    this.htmlDiv_.setAttribute('aria-labelledby', this.title_);
```
Comment (samelhusseini) [inline packages/plugins/typed-variable-modal/src/Modal.js]: This now needs to go on the modalContainer, where the role is set.

File: packages/plugins/typed-variable-modal/test/index.html [hunk 31]
```diff
@@ -0,0 +1,83 @@
+<!DOCTYPE html>
+<html>
+<head>
+    <meta charset="utf-8">
+    <title>Workspace Search Playground</title>
```
Comment (samelhusseini) [inline packages/plugins/typed-variable-modal/test/index.html]: Rename to typed variable playground.

File: packages/plugins/typed-variable-modal/src/Modal.js (hunk 30 above)
Comment (alschmiedt) [reply packages/plugins/typed-variable-modal/src/Modal.js]: updated

File: packages/plugins/typed-variable-modal/test/index.html (hunk 31 above)
Comment (alschmiedt) [reply packages/plugins/typed-variable-modal/test/index.html]: done
